src/decoder.ts

```diff
@@ -0,0 +1,595 @@
+import debug from 'debug';
+import path from 'node:path';
+import type {
+  ArduinoState,
+  BoardDetails,
+  BuildProperties,
+} from 'vscode-arduino-api';
+import { Debug, access, isWindows, neverSignal, run } from './utils';
+
+const decoderDebug: Debug = debug('espExceptionDecoder:decoder');
+
+export interface DecodeParams {
+  readonly toolPath: string;
+  readonly elfPath: string;
+  readonly fqbn: string; // UI
+  readonly sketchPath: string; // UIc
+}
+
+export async function createDecodeParams(
+  arduinoState: ArduinoState
+): Promise<DecodeParams> {
+  const { boardDetails, compileSummary, sketchPath } = arduinoState;
+  if (!sketchPath) {
+    throw new Error('Sketch path is not set');
+  }
+  const fqbn = sanitizeFqbn(arduinoState.fqbn);
+  if (!fqbn) {
+    throw new Error('No board selected');
+  }
+  const [vendor, arch] = splitFqbn(fqbn);
+  if (!boardDetails) {
+    throw new DecodeParamsError(
+      `Platform '${vendor}:${arch}' is not installed`,
+      { sketchPath, fqbn }
+    );
+  }
+  if (!supportedArchitectures.has(arch)) {
+    throw new DecodeParamsError(`Unsupported board: '${fqbn}'`, {
+      sketchPath,
+      fqbn,
+    });
+  }
+  if (!compileSummary) {
+    throw new DecodeParamsError(
+      'The summary of the previous compilation is unavailable. Compile the sketch',
+      {
+        sketchPath,
+        fqbn,
+      }
+    );
+  }
+  const { buildPath } = compileSummary;
+  const sketchFolderName = path.basename(sketchPath);
+  const [toolPath, elfPath] = await Promise.all([
+    findToolPath(boardDetails),
+    findElfPath(sketchFolderName, buildPath),
+  ]);
+  if (!elfPath) {
+    throw new DecodeParamsError(
```

Comment:
Manually pick the `.elf` file?

Reply:
I will implement it in the first iteration if there is a request. Otherwise, maybe later.

There are cases where user interaction is required by an extension tool (e.g., the ESP8266 fs tool [asks](https://github.com/esp8266/arduino-esp8266fs-plugin/blob/56a6c515545ff5301be5e679d014ab7062600d81/src/ESP8266FS.java#L288-L290) for confirmation before creating an empty image), but for the `.elf` lookup, the CLI should be sufficient to locate the file. The behavior could be unpredictable if users can pick any arbitrary elf files.

Let's see what will be the first user feedback. But I would postpone making this feature parity with the original implementation.